src/Version/SemVer.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Bump\Version;
+
+use Bump\Version;
+
+class SemVer implements Version
+{
+    /** @var string The major version */
+    private $major;
+
+    /** @var string The minor version */
+    private $minor;
+
+    /** @var string The patch version */
+    private $patch;
+
+    /** @var bool Determine if the tag is in a valid format. */
+    private $is_valid_format = false;
+
+    public function __construct($tag)
+    {
+        if ($this->init($tag)) {
+            $this->is_valid_format = true;
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function exclude()
+    {
+        return !$this->is_valid_format;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function bump($touple)
+    {
+        $touple_map = [
+            'patch' => 'bumpPatch',
+            'minor' => 'bumpMinor',
+            'major' => 'bumpMajor'
+        ];
+
+        if (!isset($touple_map[$touple])) {
+            throw new \InvalidArgumentException($touple . ' is not an allowed touple');
+        }
+
+        return $this->{$touple_map[$touple]}();
+    }
+
+    /**
+     * Generates the patch bump for the given tag.
+     *
+     * @return self
+     */
+    protected function bumpPatch()
+    {
+        $version = clone $this;
+        $version->patch++;
+        return $version;
+    }
+
+    /**
+     * Generates the minor bump for the given tag.
+     *
+     * @return string
+     */
+    protected function bumpMinor()
+    {
+        $version = clone $this;
+        $version->minor++;
+        $version->patch = 0;
+        return $version;
+    }
+
+    /**
+     * Generates the major bump for the given tag.
+     *
+     * @return string
+     */
+    protected function bumpMajor()
+    {
+        $version = clone $this;
+        $version->major++;
+        $version->patch = 0;
+        $version->minor = 0;
+        return $version;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getTag()
```

Comment:
Maybe move these public functions before `bump`, that way importance of the protected methods isn't lost.